pkg/services/ngalert/state/historian/dashboard.go

```diff
@@ -0,0 +1,89 @@
+package historian
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/patrickmn/go-cache"
+)
+
+const dashboardNotFound = int64(-1)
+
+const defaultDashboardCacheExpiry = 1 * time.Minute
+
+const minCleanupInterval = 1 * time.Second
+
+// dashboardResolver resolves dashboard UIDs to IDs with caching.
+type dashboardResolver struct {
+	dashboards dashboards.DashboardService
+	cache      *cache.Cache
+	lock       sync.Mutex
+	log        log.Logger
+}
+
+func newDashboardResolver(dbs dashboards.DashboardService, log log.Logger, expiry time.Duration) *dashboardResolver {
+	return &dashboardResolver{
+		dashboards: dbs,
+		cache:      cache.New(expiry, maxDuration(2*expiry, minCleanupInterval)),
```

Comment:
Clamping at a min duration prevents a caller from using an overly short expiry like 1ns, causing a huge number of background cache cleanups.